includes/compat/class-wc-stripe-subs-compat.php

```diff
@@ -440,28 +435,46 @@ public function add_subscription_payment_meta( $payment_meta, $subscription ) {
+	 * @since 4.0.4 Stripe source id field no longer needs to be required.
+	 * @since 4.3.1 Stripe customer id field no longer needs to be required if a customer with default card is present.
 	 * @param string $payment_method_id The ID of the payment method to validate
 	 * @param array $payment_meta associative array of meta data required for automatic payments
+	 * @param WC_Subscription $subscription The subscription object.
 	 * @return array
 	 */
-	public function validate_subscription_payment_meta( $payment_method_id, $payment_meta ) {
+	public function validate_subscription_payment_meta( $payment_method_id, $payment_meta, $subscription ) {
 		if ( $this->id === $payment_method_id ) {
 
-			if ( ! isset( $payment_meta['post_meta']['_stripe_customer_id']['value'] ) || empty( $payment_meta['post_meta']['_stripe_customer_id']['value'] ) ) {
-				throw new Exception( __( 'A "Stripe Customer ID" value is required.', 'woocommerce-gateway-stripe' ) );
-			} elseif ( 0 !== strpos( $payment_meta['post_meta']['_stripe_customer_id']['value'], 'cus_' ) ) {
+			$customer_id = isset( $payment_meta['post_meta']['_stripe_customer_id']['value'] ) && ! empty( $payment_meta['post_meta']['_stripe_customer_id']['value'] )
+				? $payment_meta['post_meta']['_stripe_customer_id']['value']
+				: null;
+
+			$source_id = isset( $payment_meta['post_meta']['_stripe_source_id']['value'] ) && ! empty( $payment_meta['post_meta']['_stripe_source_id']['value'] )
+				? $payment_meta['post_meta']['_stripe_source_id']['value']
+				: null;
+
+			// Reject if there is an invalid customer entered.
+			if ( ! empty( $customer_id ) && 0 !== strpos( $customer_id, 'cus_' ) ) {
 				throw new Exception( __( 'Invalid customer ID. A valid "Stripe Customer ID" must begin with "cus_".', 'woocommerce-gateway-stripe' ) );
 			}
 
-			if (
-				( ! empty( $payment_meta['post_meta']['_stripe_source_id']['value'] )
-				&& 0 !== strpos( $payment_meta['post_meta']['_stripe_source_id']['value'], 'card_' ) )
-				&& ( ! empty( $payment_meta['post_meta']['_stripe_source_id']['value'] )
-				&& 0 !== strpos( $payment_meta['post_meta']['_stripe_source_id']['value'], 'src_' ) ) ) {
-
+			// Reject if there is an invalid source entered.
+			if ( ! empty( $source_id ) && 0 !== strpos( $source_id, 'src_' ) && 0 !== strpos( $source_id, 'card_' ) ) {
 				throw new Exception( __( 'Invalid source ID. A valid source "Stripe Source ID" must begin with "src_" or "card_".', 'woocommerce-gateway-stripe' ) );
 			}
+
+			if ( empty( $customer_id ) ) {
+				// Notify merchants that `customer_id` is required when `source_id` was manually entered.
+				if ( ! empty( $source_id ) ) {
+					throw new Exception( __( 'A "Stripe Customer ID" value is required when "Stripe Source ID" is used.', 'woocommerce-gateway-stripe' ) );
+				}
+
+				// Before bailing, check if there is a default source that can be used.
+				$default_source = $this->get_default_customer_source_for_order( $subscription );
```

Comment:
Why do we need this call to `get_default_customer_source_for_order()`?

From what I can see, even if we find a default source, it's not possible to use it at this step because `validate_subscription_payment_meta()` is only used to validate the payment info, but it's not able to change the values that the user has already entered.

Reply:
Yes, we cannot use the source at the moment. However, we can still use it to verify whether there is an _available_ source.

This condition will check whether there is a default source (associated with the customer) only when the "Stripe Customer ID" field is empty. In that case, if a customer _is_ already chosen, and they have a default source that's okay. If they don't then the merchant should see a message (based on the text of the exception).